src/librustc/infer/error_reporting/nice_region_error/trait_impl_difference.rs

```diff
@@ -0,0 +1,56 @@
+//! Error Reporting for `impl` items that do not match the obligations from their `trait`.
+
+use syntax_pos::Span;
+use crate::ty::Ty;
+use crate::infer::{ValuePairs, Subtype};
+use crate::infer::error_reporting::nice_region_error::NiceRegionError;
+use crate::infer::lexical_region_resolve::RegionResolutionError;
+use crate::util::common::ErrorReported;
+
+impl<'a, 'tcx> NiceRegionError<'a, 'tcx> {
+    /// Print the error message for lifetime errors when the `impl` doesn't conform to the `trait`.
+    pub(super) fn try_report_impl_not_conforming_to_trait(&self) -> Option<ErrorReported> {
+        if let Some(ref error) = self.error {
+            if let RegionResolutionError::SubSupConflict(
+                _,
+                var_origin,
+                sub_origin,
+                _sub,
+                sup_origin,
+                _sup,
+            ) = error.clone() {
+                match (&sup_origin, &sub_origin) {
+                    (&Subtype(ref sup_trace), &Subtype(ref sub_trace)) => {
+                        if let (
+                            ValuePairs::Types(sub_expected_found),
+                            ValuePairs::Types(sup_expected_found),
+                        ) = (&sub_trace.values, &sup_trace.values) {
+                            if sup_expected_found == sub_expected_found {
+                                let sp = var_origin.span();
```

Comment:
Is there a way of getting the original trait item's span here? That way we could point at both to make the discrepancy clearer.

Reply:
Hmm. I think that these variables are created in/around this code here:

https://github.com/rust-lang/rust/blob/8c7b921feb5b100b1825ad54d46a8cca575b35b3/src/librustc_typeck/check/compare_method.rs#L234-L238

If that's correct, one option might be to pack more information into the "variable origin". However, that code is *also* using an obligation cause that contains all the information we need:

https://github.com/rust-lang/rust/blob/8c7b921feb5b100b1825ad54d46a8cca575b35b3/src/librustc_typeck/check/compare_method.rs#L93-L101

This is threaded to "subtype" code here:

https://github.com/rust-lang/rust/blob/8c7b921feb5b100b1825ad54d46a8cca575b35b3/src/librustc_typeck/check/compare_method.rs#L290-L294

However, that seems to get lost when converting to a subregion obligation. There is some code tries to preserve it

https://github.com/rust-lang/rust/blob/8c7b921feb5b100b1825ad54d46a8cca575b35b3/src/librustc/infer/mod.rs#L1743-L1765

so I'm not 100% sure what's going on. Would have to dump some logs.

Reply:
Should we merge the work done so far and get back to it with more time to add the extra information? The original tickets would remain open.

Reply:
The reason is because the failing obligation is of code `ReferenceOutlivesReferent`, not `CompareImplMethodObligation`. We really need to add a new variant to chain obligations together and recover the original problem.

Reply:
Scratch that, figured it out.